backend/canisters/community/impl/src/updates/register_proposal_vote.rs

```diff
@@ -0,0 +1,133 @@
+                Success
+            }
+            user_canister::c2c_vote_on_proposal::Response::NoEligibleNeurons => NoEligibleNeurons,
+            user_canister::c2c_vote_on_proposal::Response::ProposalNotFound => ProposalNotFound,
+            user_canister::c2c_vote_on_proposal::Response::ProposalNotAcceptingVotes => ProposalNotAcceptingVotes,
+            user_canister::c2c_vote_on_proposal::Response::InternalError(error) => InternalError(error),
+        },
+        Err(error) => InternalError(format!("{error:?}")),
+    }
+}
+
+struct PrepareResult {
+    user_id: UserId,
+    is_nns: bool,
+    governance_canister_id: CanisterId,
+    proposal_id: ProposalId,
+}
+
+fn prepare(args: &Args, state: &RuntimeState) -> Result<PrepareResult, Response> {
+    if state.data.is_frozen() {
+        return Err(CommunityFrozen);
+    }
+
+    let caller = state.env.caller();
+
+    let member = match state.data.members.get(caller) {
+        Some(m) => m,
+        None => return Err(UserNotInCommunity),
+    };
+
+    if member.suspended.value {
+        return Err(UserSuspended);
+    }
+
+    let channel = match state.data.channels.get(&args.channel_id) {
+        Some(c) => c,
+        None => return Err(ChannelNotFound),
+    };
+
+    let channel_member = match channel.chat.members.get(&member.user_id) {
+        Some(m) => m,
+        None => return Err(UserNotInChannel),
+    };
+
+    let min_visible_event_index = channel_member.min_visible_event_index();
+    let now = state.env.now();
+
+    if let Some(proposal) = channel
+        .chat
+        .events
+        .visible_main_events_reader(min_visible_event_index, now)
+        .message_internal(args.message_index.into())
+        .and_then(|m| if let MessageContentInternal::GovernanceProposal(p) = &m.content { Some(p) } else { None })
+    {
+        if let Some(vote) = proposal.votes.get(&member.user_id) {
+            Err(AlreadyVoted(*vote))
+        } else {
+            Ok(PrepareResult {
+                user_id: member.user_id,
+                is_nns: proposal.proposal.is_nns(),
+                governance_canister_id: proposal.governance_canister_id,
+                proposal_id: proposal.proposal.id(),
+            })
+        }
+    } else {
+        Err(ProposalMessageNotFound)
+    }
+}
+
+fn commit(channel_id: ChannelId, user_id: UserId, args: Args, state: &mut RuntimeState) -> Response {
+    let channel = match state.data.channels.get_mut(&channel_id) {
+        Some(c) => c,
+        None => return ChannelNotFound,
+    };
+
+    let member = match channel.chat.members.get_mut(&user_id) {
+        Some(m) => m,
+        None => return UserNotInChannel,
+    };
+
+    let now = state.env.now();
+    let min_visible_event_index = member.min_visible_event_index();
+
+    match channel
+        .chat
+        .events
+        .record_proposal_vote(user_id, min_visible_event_index, args.message_index, args.adopt, now)
```

Comment:
If `record_proposal_vote` was on GroupChatCore then it could push the event and update the member and avoid this being done in multiple places.

Reply:
This is temporary though right?